### PR TITLE
Print diff of object spec if validation fails during deletion

### PR DESCRIPTION
**Pull request by acumino**
**How to categorize this PR?**

/area ops-productivity
/kind enhancement

**What this PR does / why we need it**:
Previously if because of some issues shoot gets stuck in the deletion with an error that `shoot spec field is immutable`. It was hard to find which field is being modified.
This PR enhances the error to also print the particular value being changed in the spec of the object.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator
NONE
```

### CODE REVIEW

**Comment by shafeeqes** [comment]
~Can you also post how the error will look like, from our testing ?~
The errors in the tests show this perfectly.

File: pkg/apis/core/validation/shoot.go [hunk 1]
```diff
@@ -266,6 +267,9 @@ func ValidateShootSpecUpdate(newSpec, oldSpec *core.ShootSpec, newObjectMeta met
 
 	if newObjectMeta.DeletionTimestamp != nil && !apiequality.Semantic.DeepEqual(newSpec, oldSpec) {
 		allErrs = append(allErrs, apivalidation.ValidateImmutableField(newSpec, oldSpec, fldPath)...)
+		if diff := deep.Equal(newSpec, oldSpec); diff != nil {
+			allErrs = append(allErrs, field.Forbidden(fldPath, strings.Join(diff, ",")))
+		}
```
Comment (shafeeqes) [inline pkg/apis/core/validation/shoot.go]:
```suggestion
	if newObjectMeta.DeletionTimestamp != nil && !apiequality.Semantic.DeepEqual(newSpec, oldSpec) {
		errorList := apivalidation.ValidateImmutableField(newSpec, oldSpec, fldPath)
		if diff := deep.Equal(newSpec, oldSpec); diff != nil {
			errorList = field.ErrorList{field.Forbidden(fldPath, strings.Join(diff, ","))}
		}
		return errorList
	}
```
This way we can reduce duplication. The first error message doesn't help anyway. Also we can exit right away.

**Comment by shafeeqes** [comment]
There are spec comparisons in validations of extensions as well,
eg:
https://github.com/gardener/gardener/blob/2f34cb1422985253aa47eb761bb759d6ae4823e3/pkg/apis/extensions/validation/backupbucket.go#L68-L71
Also in managedseed validation, https://github.com/gardener/gardener/blob/9a8c281dd526abac2f137674e1b9ae65b3892c65/pkg/apis/seedmanagement/validation/managedseed.go#L112
Should we print the diff there as well?
We can also make this logic into a function, for eg, `ValidateSpecOnDeletion` and reuse them as well.

File: pkg/apis/core/validation/controllerinstallation.go [hunk 2]
```diff
@@ -65,8 +68,11 @@ func ValidateControllerInstallationSpecUpdate(new, old *core.ControllerInstallat
 	allErrs := field.ErrorList{}
 
 	if deletionTimestampSet && !apiequality.Semantic.DeepEqual(new, old) {
-		allErrs = append(allErrs, apivalidation.ValidateImmutableField(new, old, fldPath)...)
-		return allErrs
+		errorList := apivalidation.ValidateImmutableField(new, old, fldPath)
+		if diff := deep.Equal(new, old); diff != nil {
+			errorList = field.ErrorList{field.Forbidden(fldPath, strings.Join(diff, ","))}
+		}
+		return errorList
```
Comment (shafeeqes) [inline pkg/apis/core/validation/controllerinstallation.go]:
```suggestion
		if diff := deep.Equal(new, old); diff != nil {
			return field.ErrorList{field.Forbidden(fldPath, strings.Join(diff, ","))}
		} else {
			return apivalidation.ValidateImmutableField(new, old, fldPath)
		}
```

**Review by shafeeqes — COMMENTED** [review]
Had a combined look with @ScheererJ.
Only one comment that we can reorder this way.

**Review by shafeeqes — COMMENTED** [review]
/lgtm

**Review by ScheererJ — APPROVED** [review]
/lgtm

**Pull request merged**